src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java

```diff
@@ -1225,4 +1237,27 @@ void updateAAD(byte[] src, int offset, int len) {
         checkReinit();
         cipher.updateAAD(src, offset, len);
     }
+
+    // This must only be used with GCM.
+    // If some data has been buffered from an update call, operate on the buffer
+    // then run doFinal.
+    int doFinal(ByteBuffer src, ByteBuffer dst) throws ShortBufferException,
+        IllegalBlockSizeException, BadPaddingException {
+        int estOutSize = getOutputSizeByOperation(src.remaining(), true);
+        if (estOutSize > dst.remaining()) {
+            throw new ShortBufferException("output buffer too small");
+        }
+
+        if (decrypting) {
+            if (buffered > 0) {
+                cipher.decrypt(buffer, 0, buffered, new byte[0], 0);
+            }
+            return cipher.decryptFinal(src, dst);
+        } else {
+            if (buffered > 0) {
+                cipher.encrypt(buffer, 0, buffered, new byte[0], 0);
```

Comment:
So there is a problem here, but not a short buffer.  I hadn't realized that encrypt(byte[]..) didn't put data in the GaloisCounterMode.ibuffer because I did that with encrypt(ByteBuffer..).  I added a GCM only method that will cover this and place the data in the ibuffer for doFinal(ByteBuffer..) can complete the op.